utf8/valid_amd64.s

```diff
@@ -0,0 +1,253 @@
+	VMOVDQU cont3_vec<>+0(SB), Y2
+
+	// High nibble of current byte
+	VMOVDQU nibble1_errors<>+0(SB), Y3
+
+	// Low nibble of current byte
+	VMOVDQU nibble2_errors<>+0(SB), Y4
+
+	// High nibble of the next byte
+	VMOVDQU nibble3_errors<>+0(SB), Y5
+
+	// Nibble mask
+	VMOVDQU nibble_mask<>+0(SB), Y6
+
+	// For the first pass, set the previous block as zero.
+	VXORPS Y7, Y7, Y7
+
+	// Zeroes the error vector.
+	VXORPS Y8, Y8, Y8
+
+	// Zeroes the "previous block was incomplete" vector.
+	VXORPS Y9, Y9, Y9
+
+	// Top of the loop.
+check_input:
+	// if bytes left >= 32
+	CMPQ CX, $0x20
+
+	// go process the next block
+	JGE process
+
+	// If < 32 bytes left
+	// Fast exit if done
+	CMPQ CX, $0x00
+	JE   end
+
+	// If 0 < bytes left < 32.
+	// Prepare scratch space
+	LEAQ    (SP), BX
+	VXORPS  Y10, Y10, Y10
+	VMOVDQU Y10, (BX)
+
+	// Make a copy of the remaining bytes into the zeroed scratch space and make it the next block to read.
+	MOVQ AX, SI
+	MOVQ BX, AX
+	CMPQ CX, $0x01
+	JE   handle1
+	CMPQ CX, $0x03
+	JBE  handle2to3
+	CMPQ CX, $0x04
+	JE   handle4
+	CMPQ CX, $0x08
+	JB   handle5to7
+	JE   handle8
+	CMPQ CX, $0x10
+	JBE  handle9to16
+	CMPQ CX, $0x20
+	JBE  handle17to32
+
+handle1:
+	MOVB (SI), CL
+	MOVB CL, (AX)
+	JMP  after_copy
+
+handle2to3:
+	MOVW (SI), BX
+	MOVW -2(SI)(CX*1), SI
+	MOVW BX, (AX)
+	MOVW SI, -2(AX)(CX*1)
+	JMP  after_copy
+
+handle4:
+	MOVL (SI), CX
+	MOVL CX, (AX)
+	JMP  after_copy
+
+handle5to7:
+	MOVL (SI), BX
+	MOVL -4(SI)(CX*1), SI
+	MOVL BX, (AX)
+	MOVL SI, -4(AX)(CX*1)
+	JMP  after_copy
+
+handle8:
+	MOVQ (SI), CX
+	MOVQ CX, (AX)
+	JMP  after_copy
+
+handle9to16:
+	MOVQ (SI), BX
+	MOVQ -8(SI)(CX*1), SI
+	MOVQ BX, (AX)
+	MOVQ SI, -8(AX)(CX*1)
+	JMP  after_copy
+
+handle17to32:
+	MOVOU (SI), X10
+	MOVOU -16(SI)(CX*1), X11
+	MOVOU X10, (AX)
+	MOVOU X11, -16(AX)(CX*1)
+
+after_copy:
+	MOVQ $0x0000000000000020, CX
+
+	// Process one 32B block of data
+process:
+	// Load the next block of bytes
+	VMOVDQU (AX), Y10
+	SUBQ    $0x20, CX
+	ADDQ    $0x20, AX
+
+	// Fast check to see if ASCII
+	VPMOVMSKB Y10, BX
+	CMPL      BX, $0x00
+	JNZ       non_ascii
+
+	// If this whole block is ASCII, there is nothing to do, and it is an error if any of the previous code point was incomplete.
+	VPOR Y8, Y9, Y8
+	JMP  check_input
+
+non_ascii:
+	XORB DL, DL
+
+	// Check errors on the high nibble of the previous byte
+	VPERM2I128 $0x03, Y7, Y10, Y9
+	VPALIGNR   $0x0f, Y9, Y10, Y9
+	VPSRLW     $0x04, Y9, Y11
+	VPAND      Y11, Y6, Y11
+	VPSHUFB    Y11, Y3, Y11
+
+	// Check errors on the low nibble of the previous byte
+	VPAND   Y9, Y6, Y9
+	VPSHUFB Y9, Y4, Y9
+	VPAND   Y9, Y11, Y11
+
+	// Check errors on the high nibble on the current byte
+	VPSRLW  $0x04, Y10, Y9
+	VPAND   Y9, Y6, Y9
+	VPSHUFB Y9, Y5, Y9
+	VPAND   Y9, Y11, Y11
+
+	// Find 3 bytes continuations
+	VPERM2I128 $0x03, Y7, Y10, Y9
+	VPALIGNR   $0x0e, Y9, Y10, Y9
+	VPSUBUSB   Y2, Y9, Y9
+
+	// Find 4 bytes continuations
+	VPERM2I128 $0x03, Y7, Y10, Y7
```

Comment:
Could the result from the previous `VPERM2I128` call(s) be used here? Maybe it could be kept in a temporary register to avoid all but the first call?